### PR TITLE
Add workaround for Scope Mismatch issue for HttpClientFactory

### PR DESCRIPTION
**Pull request by CarnaViire**
## Summary

Adds workaround for accessing application DI scope for `HttpClientFactory`, as the Scope Mismatch issue https://github.com/dotnet/runtime/issues/47091 most possibly won't be fixed in .NET 8.0.

Based on workarounds suggested in the issue.

Contributes to https://github.com/dotnet/runtime/issues/47091

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/extensions/httpclient-factory.md](https://github.com/dotnet/docs/blob/1e16adc479fe5c750ce49305013de1b8b096d5dd/docs/core/extensions/httpclient-factory.md) | [IHttpClientFactory with .NET](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory?branch=pr-en-us-34864) |

### CODE REVIEW

File: docs/core/extensions/httpclient-factory.md [hunk 1]
```diff
@@ -278,12 +278,27 @@ If you need to use `HttpClient` instances in a singleton service, consider the f
 
 Users are strongly advised **not to cache scope-related information** (such as data from `HttpContext`) inside `HttpMessageHandler` instances and use scoped dependencies with caution to avoid leaking sensitive information.
 
+If you require access to an application DI scope from your message handler, e.g. for authentication, you can work around that by incapsulating scope-aware logic in a separate transient `DelegatingHandler`, and wrap it around `HttpMessageHandler` instance from the `IHttpClientFactory` cache, which you can get using `IHttpMessageHandlerFactory` by calling <xref:System.Net.Http.IHttpMessageHandlerFactory.CreateHandler%2A> for any registered _named client_. In that case, you would need to create `HttpClient` instance yourself using the constructed handler.
+
+:::image type="content" source="media/httpclientfactory-scopes-workaround.png" alt-text="Diagram showing gaining access to application DI scopes via a separate transient message handler and IHttpMessageHandlerFactory":::
+
+The following example shows creating an `HttpClient` with a scope-aware `HttpHandler`:
```
Comment (IEvangelist) [inline docs/core/extensions/httpclient-factory.md:285]:
```suggestion
The following example shows creating an `HttpClient` with a scope-aware `HttpMessageHandler`:
```

File: docs/core/extensions/httpclient-factory.md [hunk 2]
```diff
@@ -278,12 +278,27 @@ If you need to use `HttpClient` instances in a singleton service, consider the f
...
+
+:::code source="snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs" id="CreateClient":::
+
+A further workaround can follow with an extension method for registering a scope-aware `HttpHandler` and overriding default `IHttpClientFactory` registration by a transient service with access to the current application scope:
```
Comment (IEvangelist) [inline docs/core/extensions/httpclient-factory.md:289]:
```suggestion
A further workaround can follow with an extension method for registering a scope-aware `HttpMessageHandler` and overriding default `IHttpClientFactory` registration by a transient service with access to the current app scope:
```

File: docs/core/extensions/httpclient-factory.md [hunk 3]
```diff
@@ -278,12 +278,27 @@ If you need to use `HttpClient` instances in a singleton service, consider the f
 
 Users are strongly advised **not to cache scope-related information** (such as data from `HttpContext`) inside `HttpMessageHandler` instances and use scoped dependencies with caution to avoid leaking sensitive information.
 
+If you require access to an application DI scope from your message handler, e.g. for authentication, you can work around that by incapsulating scope-aware logic in a separate transient `DelegatingHandler`, and wrap it around `HttpMessageHandler` instance from the `IHttpClientFactory` cache, which you can get using `IHttpMessageHandlerFactory` by calling <xref:System.Net.Http.IHttpMessageHandlerFactory.CreateHandler%2A> for any registered _named client_. In that case, you would need to create `HttpClient` instance yourself using the constructed handler.
+
+:::image type="content" source="media/httpclientfactory-scopes-workaround.png" alt-text="Diagram showing gaining access to application DI scopes via a separate transient message handler and IHttpMessageHandlerFactory":::
```
Comment (IEvangelist) [inline docs/core/extensions/httpclient-factory.md:283]:
```suggestion
:::image type="content" source="media/httpclientfactory-scopes-workaround.png" alt-text="Diagram showing gaining access to app DI scopes via a separate transient message handler and IHttpMessageHandlerFactory":::
```

**Review by IEvangelist — COMMENTED** [review]
Looks great! I have a few nits.

File: docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        // Get or create HttpMessageHandler from HttpClientFactory
+        HttpMessageHandler handler = _httpMessageHandlerFactory.CreateHandler(name);
+
+        if (scopeAwareHandler != null)
+        {
+            scopeAwareHandler.InnerHandler = handler;
+            handler = scopeAwareHandler;
+        }
+
+        HttpClient client = new HttpClient(handler);
```
Comment (IEvangelist) [inline docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs:59]:
```suggestion
        HttpClient client = new(handler);
```

File: docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        // get custom options to get scope aware handler information
+        ScopeAwareHttpClientFactoryOptions scopeAwareOptions = _scopeAwareOptionsMonitor.Get(name);
+        Type? scopeAwareHandlerType = scopeAwareOptions.HttpHandlerType;
+
+        // <CreateClient>
+        if (scopeAwareHandlerType != null)
+        {
+            if (!typeof(DelegatingHandler).IsAssignableFrom(scopeAwareHandlerType))
+            {
+                throw new ArgumentException($"Scope aware HttpHandler {scopeAwareHandlerType.Name} should be assignable to DelegatingHandler");
```
Comment (IEvangelist) [inline docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs:38]:
```suggestion
                throw new ArgumentException($"""
                    Scope aware HttpHandler {scopeAwareHandlerType.Name} should 
                    be assignable to DelegatingHandler
                    """);
```

File: docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            }
+
+            // Create top-most delegating handler with scoped dependencies
+            scopeAwareHandler = (DelegatingHandler)_scopeServiceProvider.GetRequiredService(scopeAwareHandlerType); // should be transient
+            if (scopeAwareHandler.InnerHandler != null)
+            {
+                throw new ArgumentException($"Inner handler of a delegating handler {scopeAwareHandlerType.Name} should be null. " +
+                    "Scope aware HttpHandler should be registered as Transient.");
```
Comment (IEvangelist) [inline docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs:46]:
```suggestion
                throw new ArgumentException($"""
                    Inner handler of a delegating handler {scopeAwareHandlerType.Name} should be null. 
                    Scope aware HttpHandler should be registered as Transient.
                    """);
```

File: docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    // <AddScopeAwareHttpHandler>
+    public static IHttpClientBuilder AddScopeAwareHttpHandler<THandler>(this IHttpClientBuilder builder) where THandler : DelegatingHandler
+    {
+        builder.Services.TryAddTransient<THandler>();
+        if (!builder.Services.Any(sd => sd.ImplementationType == typeof(ScopeAwareHttpClientFactory)))
+        {
+            builder.Services.AddTransient<IHttpClientFactory, ScopeAwareHttpClientFactory>(); // override default IHttpClientFactory registration
+        }
+
+        builder.Services.Configure<ScopeAwareHttpClientFactoryOptions>(builder.Name, options => options.HttpHandlerType = typeof(THandler));
```
Comment (IEvangelist) [inline docs/core/extensions/snippets/http/scopeworkaround/ScopeAwareHttpClientFactory.cs:88]:
```suggestion
        builder.Services.Configure<ScopeAwareHttpClientFactoryOptions>(
            builder.Name, options => options.HttpHandlerType = typeof(THandler));
```

File: docs/core/extensions/httpclient-factory.md [hunk 8]
```diff
@@ -122,7 +122,7 @@ The typed client is registered as transient with DI. In the preceding code, `Add
 1. Create an instance of `TodoService`, passing in the instance of `HttpClient` to its constructor.
 
 > [!IMPORTANT]
-> Using typed clients in singleton services can be dangerous. For more information, see the [Using Typed clients in singleton services](#use-typed-clients-in-singleton-services) section.
+> Using typed clients in singleton services can be dangerous. For more information, see the [Using Typed clients in singleton services](#using-typed-clients-in-singleton-services) section.
```
[the remark this answers is not in the document]
Comment (CarnaViire) [reply docs/core/extensions/httpclient-factory.md]: Rephrased to "Avoid" because the section is about dangers and problems of such approach

**Comment by CarnaViire** [comment]
PTAL @IEvangelist

**Review by IEvangelist — APPROVED** [review]

**Pull request merged**